cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/LauncherUtils.java

```diff
@@ -404,16 +351,59 @@ private static void runBal(Path sourceRootPath, Path sourcePath, String[] args,
         ServiceLoader<LaunchListener> listeners = ServiceLoader.load(LaunchListener.class);
         listeners.forEach(listener -> listener.beforeRunProgram(runServicesOnly));
 
-        if (runServicesOnly) {
-            if (args.length > 0) {
-                throw LauncherUtils.createUsageExceptionWithHelp("arguments not allowed for services");
-            }
-            runServices(programFile);
-        } else {
-            runMain(programFile, args);
+        int exitCode;
+        try {
+            exitCode = executeCompiledProgram(programFile, args);
+        } catch (BLangUsageException | BallerinaException e) {
+            throw createUsageException(makeFirstLetterLowerCase(e.getLocalizedMessage()));
         }
+
         BLangProgramRunner.resumeStates(programFile);
         listeners.forEach(listener -> listener.afterRunProgram(runServicesOnly));
+
+        if (!programFile.isServiceEPAvailable()) {
```

Comment:
The spec says "If the main function returns an error, then the program terminates."

IINM, we are not checking if the main function returns an error here right? So even if an error is returned, if there are services, the execution continues?

Reply:
Ah yes, should exit if the exit code is not 0. Will modify this.

Reply:
Actually, the listen phase will not be executed if there is an error in the main(). But here, the problem is the exit code returned by the main() will not be passed on to the runtime when exiting.